File: 수학_소수/프로그래머스Lv0-유한소수 판별하기.py
from collections import defaultdict
def make_primes(n:int)->list:
    primes = [True] * (n + 1)
    primes[0:2] = [False, False]
    for num in range(2, int(n * 0.5) + 1):
        if primes[num]:
            for i in range(num ** 2, n + 1, num):
                primes[i] = False
    return primes

def make_prime_factorization(n):
    primes = make_primes(n)
    prime_numbers = [x for x in range(2, n + 1) if primes[x]]
    result = defaultdict(int)
    for number in prime_numbers:
        while n % number == 0 and n > 1:
            result[number] += 1
            n //= number
        if n == 1:
            break
    return result

def solution(a, b):
    molecule = make_prime_factorization(a)
    denominator = make_prime_factorization(b)
    
    # 분모 약분
    for key, value in molecule.items():
        if key in denominator:
            denominator[key] = max(0, denominator[key] - value)
    
    # 유한 소수 판별
    for key, value in denominator.items():
        if key == 2 or key == 5:
            continue
        if value > 0:
            return 2
    return 1
"""
정확성  테스트
테스트 1 〉	통과 (0.12ms, 10.1MB)
테스트 2 〉	통과 (0.19ms, 10.2MB)
테스트 3 〉	통과 (0.04ms, 10.1MB)
테스트 4 〉	통과 (0.11ms, 10.1MB)
테스트 5 〉	통과 (0.08ms, 10.2MB)
테스트 6 〉	통과 (0.07ms, 10.3MB)
테스트 7 〉	통과 (0.15ms, 10.3MB)
테스트 8 〉	통과 (0.12ms, 10.3MB)
테스트 9 〉	통과 (0.07ms, 10.4MB)
테스트 10 〉	통과 (0.26ms, 10.1MB)
테스트 11 〉	통과 (0.05ms, 10.3MB)
테스트 12 〉	통과 (0.16ms, 10MB)
테스트 13 〉	통과 (0.02ms, 10.1MB)
테스트 14 〉	통과 (0.02ms, 10.2MB)
테스트 15 〉	통과 (0.02ms, 10.2MB)
"""

File: 수학_소수/프로그래머스Lv0-유한소수 판별하기.js
function makePrimes(n) {
  const primes = Array.from({ length: n + 1 }, (v) => true)
  primes.splice(0, 2, false, false)
  for (let num = 2; num < Math.floor(Math.sqrt(n)) + 1; num++) {
    if (primes[num]) {
      for (let i = num * num; i < n + 1; i += num) {
        primes[i] = false
      }
    }
  }
  return primes
}

function makePrimeFactorization(n) {
  const primes = makePrimes(n)
  const primeNumbers = Array.from({ length: n + 1 }, (v, i) => i).filter((x) => primes[x])
  const result = {}
  for (const number of primeNumbers) {
    if (n % number === 0) {
      result[number] = 0
    }
    while (n % number === 0 && n > 1) {
      result[number]++
      n /= number
    }
    if (n === 1) break
  }
  return result
}

function solution(a, b) {
  const numerator = makePrimeFactorization(a)
  const denominator = makePrimeFactorization(b)

  for (const [key, value] of Object.entries(numerator)) {
    if (key in denominator) {
      denominator[key] = Math.max(0, denominator[key] - value)
    }
  }

  for (const [key, value] of Object.entries(denominator)) {
    if (key === '2' || key === '5') continue
    if (value > 0) return 2
  }
  return 1
}
// 정확성  테스트
// 테스트 1 〉	통과 (0.43ms, 33.2MB)
// 테스트 2 〉	통과 (0.55ms, 33.4MB)
// 테스트 3 〉	통과 (0.33ms, 33.5MB)
// 테스트 4 〉	통과 (0.39ms, 33.4MB)
// 테스트 5 〉	통과 (0.41ms, 33.4MB)
// 테스트 6 〉	통과 (0.36ms, 33.5MB)
// 테스트 7 〉	통과 (0.52ms, 33.5MB)
// 테스트 8 〉	통과 (0.47ms, 33.6MB)
// 테스트 9 〉	통과 (0.37ms, 33.4MB)
// 테스트 10 〉	통과 (0.66ms, 33.5MB)
// 테스트 11 〉	통과 (0.35ms, 33.4MB)
// 테스트 12 〉	통과 (0.48ms, 33.4MB)
// 테스트 13 〉	통과 (0.28ms, 33.4MB)
// 테스트 14 〉	통과 (0.30ms, 33.4MB)
// 테스트 15 〉	통과 (0.30ms, 33.5MB)
